src/main/src/main/java/org/geoserver/ows/ProxifyingURLMangler.java

```diff
@@ -5,33 +5,129 @@
+            return header;
+        }
+    }
+
     GeoServer geoServer;
 
+    public static String TEMPLATE_SEPARATOR = " ";
+    public static String TEMPLATE_PREFIX = "${";
+    public static String TEMPLATE_POSTFIX = "}";
+
     public ProxifyingURLMangler(GeoServer geoServer) {
         this.geoServer = geoServer;
     }
 
     public void mangleURL(
             StringBuilder baseURL, StringBuilder path, Map<String, String> kvp, URLType type) {
 
-        // first check the system property
-        String proxyBase = GeoServerExtensions.getProperty("PROXY_BASE_URL");
-        if (proxyBase == null) {
-            // if no system property fall back to configuration
-            proxyBase = geoServer.getSettings().getProxyBaseUrl();
-        }
+        // first check the system property, then fall back to configuration
+        String proxyBase =
+                (GeoServerExtensions.getProperty(Requests.PROXY_PARAM) != null)
+                        ? GeoServerExtensions.getProperty(Requests.PROXY_PARAM)
+                        : this.geoServer.getSettings().getProxyBaseUrl();
+
+        // Mangles the URL base in different ways based on a flag
+        // (for two reasons: a) speed; b) to make the admin aware of
+        // possible security liabilities)
+        baseURL =
+                (this.geoServer.getGlobal().isUseHeadersProxyURL() == true && proxyBase != null)
+                        ? this.mangleURLHeaders(baseURL, proxyBase)
+                        : this.mangleURLFixedURL(baseURL, proxyBase);
+    }
+
+    /**
+     * Mangle URL using the PROXY_BASE_URL
+     *
+     * @param baseURL URL to mangle
+     * @param proxyBase Proxy base URL as taken from configuration
+     * @return mangled proxy URL
+     */
+    private StringBuilder mangleURLFixedURL(StringBuilder baseURL, String proxyBase) {
 
-        // perform the replacement if the proxy base is set
+        // perform the replacement if the proxy base is set,
+        // otherwise return the baseURL unchanged
         if (proxyBase != null && proxyBase.trim().length() > 0) {
             baseURL.setLength(0);
             baseURL.append(proxyBase);
         }
+
+        return baseURL;
+    }
+
+    /**
+     * Mangle URL using request headers
+     *
+     * @param baseURL URL to mangle
+     * @param proxyBase Proxy base URL as taken from configuration
+     * @return mangled proxy URL
+     */
+    private StringBuilder mangleURLHeaders(StringBuilder baseURL, String proxyBase) {
+
+        // If the proxy base URL does not contain templates, fall back to
+        // the fixed URL cse
+        if (!proxyBase.contains(TEMPLATE_PREFIX)) {
+            return this.mangleURLFixedURL(baseURL, proxyBase);
+        }
+
+        // Cycles through templates in the proxy base URL until one is competely matched
+        Map<String, String> headers = this.compileHeadersMap();
+        for (String template : Arrays.asList(proxyBase.split(TEMPLATE_SEPARATOR))) {
```

Comment:
Did you consider using a `Pattern` to implement template expansion? See `InterpolationProperties` in gs-app-schema.

Note to GeoTools developers: we should really move `InterpolationProperties` to gt-metadata for wider use as this is the second thing I have seen this year that could have used it.

Reply:
The implementation was meant, as described in the javadoc, to be as fast as possible (Freemarker was too slow for the job, given that we potentially have a different template for each record). I did not know about InterpolationProperties, performance should be comparable (relative to large Freemarker init time for new templates, at the very least).

Reply:
IIUC InterpolationProperties is case-sensitive in its substitutions, which clashes with the case-in-sensitiveness of HTTP headers: what about adding this option to it?